### PR TITLE
BUG: get_indexer returned dtype

### LINKED ISSUE #36359: BUG: get_indexer methods return int64 instead of intp arrays
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [x] (optional) I have confirmed this bug exists on the master branch of pandas.

---
#### Code Sample, a copy-pastable example

```python
>>> import pandas as pd
>>> ax1 = pd.Index([1, 2, 3])
>>> ax2 = pd.Index([1, 1, 2])
>>> ans1 = ax1.get_indexer([1])
>>> ans2 = ax2.get_indexer_non_unique([1])
>>> print(ans1, ans1.dtype)
[0] int64
>>> print(ans2[0], ans2[0].dtype, ans2[1], ans2[1].dtype)
[0 1] int64 [] int64
```

#### Problem description
Found in #35498.  When looking at the implementation of the `get_indexer` or `get_indexer_non_unique` in pandas/_libs/index.pyx, I noticed that the returned array dtype will always be int64. Since these methods return indices arrays, I believe that intp is a more appropriate type because it will choose a size depending on ssize_t, which is guaranteed to be large enough to represent all possible indices in the array.

#### Expected Output

```python
[0] intp
[0 1] intp [] intp
```

#### Output of ``pd.show_versions()``

```python
>>> pd.show_versions()

INSTALLED VERSIONS
------------------
commit           : 7b14cf6b0b9dbcddce7b9bb22a81c73bdebc1be8
python           : 3.7.7.final.0
python-bits      : 64
OS               : Linux
OS-release       : 4.19.76-linuxkit
Version          : #1 SMP Tue May 26 11:42:35 UTC 2020
machine          : x86_64
processor        : 
byteorder        : little
LC_ALL           : C.UTF-8
LANG             : C.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.1.0rc0+406.g7b14cf6b0
numpy            : 1.18.5
pytz             : 2020.1
dateutil         : 2.8.1
pip              : 20.1.1
setuptools       : 45.2.0.post20200210
Cython           : 0.29.21
pytest           : 5.4.3
hypothesis       : 5.20.2
sphinx           : 3.1.1
blosc            : None
feather          : None
xlsxwriter       : 1.2.9
lxml.etree       : 4.4.1
html5lib         : 1.1
pymysql          : None
psycopg2         : None
jinja2           : 2.11.2
IPython          : 7.16.1
pandas_datareader: None
bs4              : 4.9.1
bottleneck       : 1.3.2
fsspec           : 0.7.4
fastparquet      : 0.4.1
gcsfs            : 0.6.2
matplotlib       : 3.2.1
numexpr          : 2.7.1
odfpy            : None
openpyxl         : 3.0.4
pandas_gbq       : None
pyarrow          : 0.16.0
pytables         : None
pyxlsb           : None
s3fs             : 0.4.2
scipy            : 1.5.1
sqlalchemy       : 1.3.18
tables           : 3.6.1
tabulate         : 0.8.7
xarray           : 0.16.0
xlrd             : 1.2.0
xlwt             : 1.3.0
numba            : 0.50.1
```

### PR DESCRIPTION
**Pull request by alexhlim**
- [x] closes #36359
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Changed the return dtype of IndexEngine’s `get_indexer()` and `get_indexer_non_unique()`. I decided to not include tests because dtypes are already implicitly tested in the following:

- pandas/tests/indexes/test_base.py (TestIndex.test_get_indexer)
- pandas/tests/base/test_misc.py (test_get_indexer_non_unique_dtype_mismatch)

Also, I ran the asv benchmarks:

```
asv continuous -f 1.1 upstream/master HEAD -b ^indexing_engines
BENCHMARKS NOT SIGNIFICANTLY CHANGED.
```

### CODE REVIEW

File: pandas/_libs/index.pyx [hunk 1]
```diff
@@ -255,7 +255,7 @@ cdef class IndexEngine:
 
     def get_indexer(self, values):
         self._ensure_mapping_populated()
-        return self.mapping.lookup(values)
+        return self.mapping.lookup(values).astype('intp')
```
Comment (jreback) [inline pandas/_libs/index.pyx:258]: hmm we don't want to astype, rather change the underlying

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/_libs/index.pyx (hunk 1 above)
Comment (alexhlim) [reply pandas/_libs/index.pyx]: gotcha, changed the return values in HashTable below (failure unrelated)

File: pandas/_libs/hashtable_class_helper.pxi.in [hunk 2]
```diff
@@ -347,7 +347,7 @@ cdef class {{name}}HashTable(HashTable):
             int ret = 0
             {{dtype}}_t val
             khiter_t k
-            int64_t[:] locs = np.empty(n, dtype=np.int64)
```
Comment (jreback) [inline pandas/_libs/hashtable_class_helper.pxi.in:350]: are these all the int64's in this file? (that are indexers)

File: pandas/_libs/index.pyx [hunk 3]
```diff
@@ -266,7 +266,7 @@ cdef class IndexEngine:
         """
         cdef:
             ndarray values, x
-            ndarray[int64_t] result, missing
```
Comment (jreback) [inline pandas/_libs/index.pyx:269]: are all of these indexer changed int his file?

**Review by jreback — CHANGES_REQUESTED** [review]
pls merge master as well

**Comment by jreback** [comment]
cc @jbrockmendel

**Comment by jbrockmendel** [comment]
LGTM

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @alexhlim

**Pull request merged**